bindings/rust/evmc-declare-tests/src/lib.rs

```diff
@@ -16,7 +16,7 @@ impl EvmcVm for FooVM {
         FooVM {}
     }
 
-    fn execute(&self, _code: &[u8], _context: &ExecutionContext) -> ExecutionResult {
-        ExecutionResult::success(1337, None)
+    fn execute(&self, _code: &[u8], _context: &ExecutionContext) -> Result<ExecutionResult, ()> {
```

Comment:
Idea: Instead of having `()` as the err type, introduce a trait which the handler (the macro-generated code) uses to interpret the error type as an EVMC exit code. Add a trait-bounded associated type to the `EvmcVM` trait for the Error type parameter of this function's return.
That way, as long as the user implements the `EvmcError` trait on a type, that type can be used for more sophisticated error handling in a way that is still compatible with the EVMC API.